Translate `_d_newarray{U,iT,T}` to a single template

teodutu opened the pull request:
This achieves the following:
- Convert `_d_newarray{U,iT,T}` to a single template `_d_newarrayT` that handles arrays of elements that either have an init symbol or are zero-initialised.
- Move compiler lowering to the semantic phase
- Store lowered expression in `NewExp.lowering`
- Copy or move array utils functions as templates from `rt.lifetime.d` to `core.internal.array.utils.d`. Some cannot be removed from `rt.lifetime.d` because they're still used by non-template hooks and it's impossible to get a type from a `TypeInfo` argument at compile time as required by templates.

This is still a draft because some tests are failing locally and I don't see how they're related to my changes. I just wanted to see if they also fail when running the tests remotely.

ibuclaw commented:
Is `OVERRIDE_MEMALLOC` being considered here? I notice that newitem and newclass already seem to be dead. Potentially at the cost of compile-time.

teodutu commented:
> `OVERRIDE_MEMALLOC`

@ibuclaw, I will take care of `_d_newclass` and `_d_newitemT` in the future but don't see any `_d_newarray*` hook defined under `OVERRIDE_MALLOC` here [1]. Is there any other place?

[1] https://github.com/dlang/dmd/blob/646fc3d74c4c065ffdc95c2260ea3390a7de42b0/compiler/src/dmd/root/rmem.d#L208-L279

teodutu commented:
So far the DMD and phobos tests are passing. `core/internal/array/duplication.d` calls `_d_newarrayU`. When I tried to split `_d_newarrayT` into `_d_newarrayU` that only allocates memory and `_d_newarrayT` that initialises it, I broke some stuff in the GC. That's why there's commented code added in the PR.

The builkite failure seems unrelated. I tested it locally and it gives the same error when using either `dmd-2.104.1` or my branch. Can you please confirm this is true? If so, I'll try to fix the allocation errors I mentioned above and update `core/internal/array/duplication.d` to use the new template as well.

teodutu committed:
druntime: Copy array allocation functions to `core.internal.array.utils`

This copies `__setArrayAllocLength()`, `__arrayAlloc()` and moves
`__arrayStart()` and `__arrayClearPad()` to
`core.internal.array.utils.d`. This is needed because `_d_newarrayT()`
calls these functions from `rt.lifetime.d`, but the file cannot be
imported from `core.internal.array.creation.d`.

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu committed:
Translate `_d_newarray{U,iT,T}` to a single template

This achieves the following:
- Convert `_d_newarray{U,iT,T}` to a single template `_d_newarrayT` that
handles arrays of elements that either have an init symbol or are
zero-initialised.
- Move compiler lowering to the semantic phase
- Store lowered expression in `NewExp.lowering`

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu commented:
`dup()` still uses the old hook [1]. I'll leave those for a later PR as this one is already large enough as it is. In addition, getting this merged will allow me to start working on `_d_arraym{iT,T}X` as well [2].

[1] https://github.com/dlang/dmd/blob/396fc8d477ca66e20f6b4fec4d532f2aee053ba3/druntime/src/core/internal/array/duplication.d#L11
[2] https://github.com/dlang/dmd/blob/396fc8d477ca66e20f6b4fec4d532f2aee053ba3/druntime/src/rt/lifetime.d#L1135C19-L1158

RazvanN7 commented:
Up until this point we have not put into changelogs the templatization of druntime hooks. However, if people find it useful we could add it. @WalterBright @ibuclaw what do you think?

teodutu commented:
@ryuukk, @RazvanN7, I created PR #15729 with a changelog entry about this. We can discuss this further there.

schveiguy commented:
@RazvanN7 Having a changelog helps knowing what you have to do for each version of the compiler you want to support with your custom runtime.
